src/Shared/FileMatcher.cs

```diff
@@ -2042,18 +2076,45 @@ enum SearchAction
                 return SearchAction.ReturnEmptyList;
             }
 
+            string directoryPattern = null;
+            if (wildcardDirectoryPart.Length > 0)
+            {
+                // If the wildcard directory part looks like "**/{pattern}/**", we are essentially looking for files that have
+                // a matching directory anywhere on their path. This is commonly used when excluding hidden directories using
+                // "**/.*/**" for example, and is worth special-casing so it doesn't fall into the slow regex logic.
+                string wildcard = wildcardDirectoryPart.TrimTrailingSlashes();
+                int wildcardLength = wildcard.Length;
+                if (wildcardLength > 6 &&
+                    wildcard[0] == '*' &&
+                    wildcard[1] == '*' &&
+                    FileUtilities.IsAnySlash(wildcard[2]) &&
+                    FileUtilities.IsAnySlash(wildcard[wildcardLength - 3]) &&
+                    wildcard[wildcardLength - 2] == '*' &&
+                    wildcard[wildcardLength - 1] == '*')
+                {
+                    // Check that there are no other slashes in the wildcard.
+                    if (wildcard.IndexOfAny(FileUtilities.Slashes, 3, wildcardLength - 6) == -1)
```

Comment:
I'm also wondering how common the case you're looking at here is. My gut tells me it's less common than the last case you considered. Also, could we allow slashes in the interior if we also fix them to be a common form and disallow wild cards?

Reply:
In .NET SDK, this is the `**/.*/**` pattern used for excluding files from hidden directories, such as `.git`. If we allow slashes in the interior, it's no longer going to be "do we have a directory on the path matching {some pattern}?" and things get more complicated because we can't simply check a directory when we're descending into it. I'm actually fine letting this fall into the slow regex case because it's not used in the SDK.